src/geometry/manifold/diffeomorph.lean

```diff
@@ -39,64 +53,365 @@ variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+⟨λ Hhf, by simpa only [(∘), h.symm_apply_apply]
+  using (h.symm.times_cont_mdiff_at.of_le hm).comp_times_cont_mdiff_within_at _ Hhf,
+  λ Hf, (h.times_cont_mdiff_at.of_le hm).comp_times_cont_mdiff_within_at _ Hf⟩
+
+@[simp] lemma diffeomorph_comp_times_cont_mdiff_at_iff {m} (h : M ≃ₘ^n⟮I, J⟯ N) {f : M' → M}
+  (hm : m ≤ n) {x} :
+  times_cont_mdiff_at I' J m (h ∘ f) x ↔ times_cont_mdiff_at I' I m f x :=
+h.diffeomorph_comp_times_cont_mdiff_within_at_iff hm
+
+@[simp] lemma diffeomorph_comp_times_cont_mdiff_on_iff {m} (h : M ≃ₘ^n⟮I, J⟯ N) {f : M' → M}
+  (hm : m ≤ n) {s} :
+  times_cont_mdiff_on I' J m (h ∘ f) s ↔ times_cont_mdiff_on I' I m f s :=
+forall_congr $ λ x, forall_congr $ λ hx, h.diffeomorph_comp_times_cont_mdiff_within_at_iff hm
+
+@[simp] lemma diffeomorph_comp_times_cont_mdiff_iff {m} (h : M ≃ₘ^n⟮I, J⟯ N) {f : M' → M}
+  (hm : m ≤ n) :
+  times_cont_mdiff I' J m (h ∘ f) ↔ times_cont_mdiff I' I m f :=
+forall_congr $ λ x, h.diffeomorph_comp_times_cont_mdiff_within_at_iff hm
+
+lemma to_local_homeomorph_mdifferentiable (h : M ≃ₘ^n⟮I, J⟯ N) (hn : 1 ≤ n) :
+  h.to_homeomorph.to_local_homeomorph.mdifferentiable I J :=
+⟨h.mdifferentiable_on _ hn, h.symm.mdifferentiable_on _ hn⟩
+
+variables [smooth_manifold_with_corners I M] [smooth_manifold_with_corners J N]
+
+lemma unique_mdiff_on_image_aux (h : M ≃ₘ^n⟮I, J⟯ N) (hn : 1 ≤ n)
+  {s : set M} (hs : unique_mdiff_on I s) :
+  unique_mdiff_on J (h '' s) :=
+begin
+  convert hs.unique_mdiff_on_preimage (h.to_local_homeomorph_mdifferentiable hn),
+  simp [h.image_eq_preimage]
+end
+
+@[simp] lemma unique_mdiff_on_image (h : M ≃ₘ^n⟮I, J⟯ N) (hn : 1 ≤ n) {s : set M} :
+  unique_mdiff_on J (h '' s) ↔ unique_mdiff_on I s :=
+⟨λ hs, h.symm_image_image s ▸ h.symm.unique_mdiff_on_image_aux hn hs,
+  h.unique_mdiff_on_image_aux hn⟩
+
+@[simp] lemma unique_mdiff_on_preimage (h : M ≃ₘ^n⟮I, J⟯ N) (hn : 1 ≤ n) {s : set N} :
+  unique_mdiff_on I (h ⁻¹' s) ↔ unique_mdiff_on J s :=
+h.symm_image_eq_preimage s ▸ h.symm.unique_mdiff_on_image hn
+
+@[simp] lemma unique_diff_on_image (h : E ≃ₘ^n[𝕜] F) (hn : 1 ≤ n) {s : set E} :
+  unique_diff_on 𝕜 (h '' s) ↔ unique_diff_on 𝕜 s :=
+by simp only [← unique_mdiff_on_iff_unique_diff_on, unique_mdiff_on_image, hn]
+
+@[simp] lemma unique_diff_on_preimage (h : E ≃ₘ^n[𝕜] F) (hn : 1 ≤ n) {s : set F} :
+  unique_diff_on 𝕜 (h ⁻¹' s) ↔ unique_diff_on 𝕜 s :=
+h.symm_image_eq_preimage s ▸ h.symm.unique_diff_on_image hn
+
+end diffeomorph
+
+namespace continuous_linear_equiv
+
+variable (e : E ≃L[𝕜] E')
+
+/-- A continuous linear equivalence between normed spaces is a diffeomorphism. -/
+def to_diffeomorph : E ≃ₘ[𝕜] E' :=
+{ times_cont_mdiff_to_fun := e.times_cont_diff.times_cont_mdiff,
+  times_cont_mdiff_inv_fun := e.symm.times_cont_diff.times_cont_mdiff,
+  to_equiv := e.to_linear_equiv.to_equiv }
+
+@[simp] lemma coe_to_diffeomorph : ⇑e.to_diffeomorph = e := rfl
+@[simp] lemma symm_to_diffeomorph : e.to_diffeomorph.symm = e.symm.to_diffeomorph := rfl
```

Comment:
Would you agree to swap sides on this one? :-) https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/simp.20normal.20form.20for.20.60e.2Eto_equiv.2Esymm.60/near/228641270

Reply:
Done. Thanks!